### PR TITLE
Small research releated refactor

### PR DESCRIPTION
**Pull request by tupaschoal**
Might be clearer to look commit by commit
- Uses `ResearchCategory` as enum throughout
- Moves Research related string stuff into `ResearchItem`
- Add converter from `RideCategory` to `ResearchCategory` to reduce the number of casts (and prevent a possible number mismatch, in case those are ever reordered)

### CODE REVIEW

**Review by duncanspumpkin — APPROVED** [review]

File: src/openrct2/ride/RideData.cpp [hunk 1]
```diff
@@ -329,3 +330,26 @@ bool RideTypeDescriptor::SupportsTrackPiece(const uint64_t trackPiece) const
 {
     return GetAvailableTrackPieces() & (1ULL << trackPiece);
 }
+
+ResearchCategory RideTypeDescriptor::RideToResearchCategory() const
+{
```
Comment (Gymnasiast) [inline src/openrct2/ride/RideData.cpp:335]: Call this GetResearchCategory()

**Comment by Gymnasiast** [comment]
What is the performance of using an unordered_map compared to an array?

**Comment by AaronVanGeffen** [comment]
> What is the performance of using an unordered_map compared to an array?

According to [cppreference](https://en.cppreference.com/w/cpp/container/unordered_map) (emphasis mine):
> Search, insertion, and removal of elements have average constant-time complexity. 

So that should be fine.

**Comment by IntelOrca** [comment]
It still won't be as fast as an array. It would be nice if we had a data structure which used the same implementation as a switch. Have a 1:1 lookup table within the bounds of the lowest and highest value.

**Comment by duncanspumpkin** [comment]
Could do with the same structure for openloco

**Comment by Gymnasiast** [comment]
Considering @IntelOrca 's comment, I think that an array would be more appropriate. After all, the source values are all adjacent.

**Comment by tupaschoal** [comment]
It's true that though the map is supposed to be constant time, it will be slower than the array. I want to prevent us from relying on the ordering of the enum. If it ever receives another value or it's reordered, the result will be completely wrong. I'd vouch for the `switch...case`, according to quick bench it has the same performance as the array (actually a bit faster):
https://quick-bench.com/q/Jwwxqi2Gq1MAUwNAkP5MYRGCZAQ
image

**Comment by tupaschoal** [comment]
Changed to a switch based on the result above

**Comment by ZehMatt** [comment]
I'm curious to how a switch is faster than the array. switch tables usually have some code around them for bounds checking and based on the size of each case branch it may cause cache misses, strange stuff.

**Comment by tupaschoal** [comment]
Would you still be ok with the switch in here anyway?

**Comment by Gymnasiast** [comment]
I still prefer the array, but I'll let @ZehMatt decide.

**Comment by tupaschoal** [comment]
Note that there are benefits to both array and switch, namely speed and "bound checking" respectively

**Comment by ZehMatt** [comment]
I would like the array honestly but its not an absolute must, if you can avoid code you should, a switch is a bit different compared to having it as pure data in the array. The complexity of arrays is always linear, switches do not promise anything.

**Comment by IntelOrca** [comment]
If this were a language like Python or F#, a switch / match would be much more elegant, and I would prefer that over an array because then it can't break if you change the order of the enum values.

**Comment by Gymnasiast** [comment]
OK, it's time to take a decision. Since @ZehMatt and I both prefer an array, let's just go for that - an array with a bounds check to avoid reading invalid memory.

**Comment by tupaschoal** [comment]
Restored the arrays, should be all good now @Gymnasiast

**Pull request merged**